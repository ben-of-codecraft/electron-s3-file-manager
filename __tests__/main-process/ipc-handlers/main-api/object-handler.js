const https = require('https');
const path = require('path');
const {
	api,
} = require('../../../../src/main-process/preload/utils');
const {
	AWS_CONFIG,
	mockIpcMainApi,
	setS3Settings,
	cleanDatabase,
} = require('../../utils');

let image;

beforeAll(async () => {
	mockIpcMainApi();
	await setS3Settings();
});

afterAll(async () => {
	await cleanDatabase();
});

describe('ipc main api object handler', () => {
	test.concurrent('sync objects from s3', async () => {
		await api.syncObjectsFromS3();
	});

	test.concurrent('delete all objects', async () => {
		const objects = await api.getObjects({limit: 500});
		const objectIds = objects.items.map(object => object.id);

		await api.deleteObjects({ids: objectIds});
	});

	test.concurrent('create folder', async () => {
		const object = await api.createFolder({basename: 'folder'});

		expect(object).toMatchSnapshot({
			id: expect.any(Number),
			createdAt: expect.any(Date),
			updatedAt: expect.any(Date),
		});
	});

	test.concurrent('create image file', async () => {
		const onProgress = jest.fn();

		image = await api.createFile({
			onProgress,
			localPath: path.join(__dirname, '..', '..', 'resources', 'image.png'),
			dirname: 'folder',
		});

		expect(image).toMatchSnapshot({
			id: expect.any(Number),
			createdAt: expect.any(Date),
			updatedAt: expect.any(Date),
			lastModified: expect.any(Date),
		});
		expect(onProgress).toBeCalledWith(
			expect.objectContaining({}),
			{
				Bucket: AWS_CONFIG.bucket,
				Key: 'folder/image.png',
				loaded: 107,
				part: 1,
				total: 107,
			},
		);
	});

	test.concurrent('get image file', async () => {
		const object = await api.getObject({id: image.id});

		expect(object).toMatchSnapshot({
			id: expect.any(Number),
			url: expect.any(String),
			createdAt: expect.any(Date),
			updatedAt: expect.any(Date),
			lastModified: expect.any(Date),
			objectHeaders: {
				$metadata: {
					extendedRequestId: expect.any(String),
				},
				Expiration: expect.any(String),
				LastModified: expect.any(Date),
			},
		});

		await new Promise((resolve, reject) => {
			const req = https.get(object.url, response => {
				expect(response.statusCode).toEqual(200);
				expect(response.headers).toMatchSnapshot({
					date: expect.any(String),
					'last-modified': expect.any(String),
					'x-amz-expiration': expect.any(String),
					'x-amz-id-2': expect.any(String),
					'x-amz-request-id': expect.any(String),
				});
				resolve();
			});

			req.on('error', error => {
				reject(error);
			});
		});
	});
});
