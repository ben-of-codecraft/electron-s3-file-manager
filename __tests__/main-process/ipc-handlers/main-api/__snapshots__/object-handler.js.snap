// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ipc main api object handler create folder 1`] = `
{
  "basename": "folder",
  "createdAt": Any<Date>,
  "dirname": "",
  "id": Any<Number>,
  "path": "folder/",
  "type": 1,
  "updatedAt": Any<Date>,
}
`;

exports[`ipc main api object handler create image file 1`] = `
{
  "basename": "image.png",
  "createdAt": Any<Date>,
  "dirname": "folder",
  "id": Any<Number>,
  "lastModified": Any<Date>,
  "path": "folder/image.png",
  "size": 107,
  "storageClass": "STANDARD",
  "type": 2,
  "updatedAt": Any<Date>,
}
`;

exports[`ipc main api object handler get image file 1`] = `
{
  "basename": "image.png",
  "createdAt": Any<Date>,
  "dirname": "folder",
  "id": Any<Number>,
  "lastModified": Any<Date>,
  "objectHeaders": {
    "$metadata": {
      "attempts": 1,
      "cfId": undefined,
      "extendedRequestId": Any<String>,
      "httpStatusCode": 200,
      "requestId": Any<String>,
      "totalRetryDelay": 0,
    },
    "AcceptRanges": "bytes",
    "ContentLength": 107,
    "ContentType": "image/png",
    "ETag": ""9514e829532c92987e4ed2294153480e"",
    "Expiration": Any<String>,
    "LastModified": Any<Date>,
    "Metadata": {},
  },
  "path": "folder/image.png",
  "size": 107,
  "storageClass": "STANDARD",
  "type": 2,
  "updatedAt": Any<Date>,
  "url": Any<String>,
}
`;

exports[`ipc main api object handler get image file 2`] = `
{
  "accept-ranges": "bytes",
  "connection": "close",
  "content-length": "107",
  "content-type": "image/png",
  "date": Any<String>,
  "etag": ""9514e829532c92987e4ed2294153480e"",
  "last-modified": Any<String>,
  "server": "AmazonS3",
  "x-amz-expiration": Any<String>,
  "x-amz-id-2": Any<String>,
  "x-amz-request-id": Any<String>,
}
`;
